torch/onnx/_internal/fx/passes/modularization.py

```diff
@@ -0,0 +1,648 @@
+                    # popped from the stack for this 'getitem' node.
+                    getitem_node.meta["nn_module_stack"] = copy.copy(
+                        ref_output.meta["nn_module_stack"]
+                    )
+                    # The node is associated with the parent module.
+                    getitem_node.meta["nn_module_stack"].popitem()
+                    copy_env[ref_output] = getitem_node
+            else:
+                # Module node has single output. Use module node directly.
+                copy_env[ref_submodule_outputs[0]] = submodule_node
+                submodule_node.meta = copy.copy(ref_submodule_outputs[0].meta)
+
+            # Update meta for new call_module node.
+            if (stack_trace := ir_node.stack_trace) is not None:
+                submodule_node.meta["stack_trace"] = stack_trace
+            raw_module_stack_meta = ir_node.stack_meta.raw_meta
+            assert raw_module_stack_meta is not None
+            submodule_node.meta["nn_module_stack"] = copy.copy(raw_module_stack_meta)
+            # The node is associated with the parent module.
+            submodule_node.meta["nn_module_stack"].popitem()
+
+        new_nodes = fx_graph.nodes
+        # Skip if the last node is already 'output'. This is the case for root module.
+        # Otherwise create an 'output' node for the inferred outputs.
+        if next(iter(reversed(new_nodes))).op != "output":
+            ref_submodule_outputs = self.module_outputs()
+            new_outputs = [copy_env[ref_output] for ref_output in self.module_outputs()]
+            node = fx_graph.output(
+                new_outputs[0] if len(new_outputs) == 1 else new_outputs
+            )
+
+        return torch.fx.GraphModule(self._reference_module, fx_graph, module_class_name)
+
+
+class _LeafNode(_IRNode):
+    """Representing a single fx.Node."""
+
+    def __init__(self, node: torch.fx.Node):
+        self._node = node
+        self._stack_meta = _module_stack_meta_from_node(node)
+
+    @property
+    def fx_op(self) -> str:
+        """Syntax sugar for self.fx_node.op."""
+        return self._node.op
+
+    @property
+    def fx_node(self) -> torch.fx.Node:
+        """Returns the fx.Node this instance represents."""
+        return self._node
+
+    @property
+    def stack_meta(self) -> _ModuleStackMeta:
+        """Returns the module stack meta data associated with this node."""
+        return self._stack_meta
+
+    @property
+    def stack_trace(self) -> Optional[str]:
+        """Returns the stack trace associated with this node."""
+        return self.fx_node.meta.get("stack_trace")
+
+    def __str__(self) -> str:
+        return f"LeafNode({self._node})"
+
+
+class Modularize(_pass.Transform):
+    """Transforms a flattened fx.GraphModule into a modular structure.
+
+    This method generates a new fx.GraphModule by creating fx submodules for all nodes
```

Comment:
Kind of get lost in the whole paragraph. I thought fx submodules was sub fx.GraphModules, but now it seems like it's something different.

Looks like fx.GraphModule flatten the nested nn.Module in fx graph with logging them with node of `call_module`, and this pass is bringing back the nested structure to facilitate module as function exporting.

Is `call_module` only invokes by this pass now? I guess my question is that is `call_module` equal to `export module as function`?

Reply:
`call_module` is non-existent in previous fx graphs, (not considering fx cond). This pass creates `call_module` nodes, and the the corresponding sub `fx.GraphModule` as the node's target.

So in terms of relation with onnx, `call_module` node is converted to the function node in ONNX, and sub `fx.GraphModule` is converted to the function proto. This behavior is documented under `def call_module` in `fx_onnx_interpreter.py`.